impedance/models/circuits/fitting.py

```diff
@@ -66,37 +86,67 @@ def circuit_fit(frequencies, impedances, circuit, initial_guess, constants,
     Currently, an error of -1 is returned.
 
     """
-    f = frequencies
     Z = impedances
 
     # extract the elements from the circuit
     extracted_elements = extract_circuit_elements(circuit)
 
     # set upper and lower bounds on a per-element basis
     if bounds is None:
-        lb, ub = [], []
+        lower_bounds, upper_bounds = [], []
         for elem in extracted_elements:
             raw_element = get_element_from_name(elem)
             for i in range(check_and_eval(raw_element).num_params):
                 if elem in constants or elem + '_{}'.format(i) in constants:
                     continue
                 if raw_element in ['CPE', 'La'] and i == 1:
-                    ub.append(1)
+                    upper_bounds.append(1)
                 else:
-                    ub.append(np.inf)
-                lb.append(0)
-        bounds = ((lb), (ub))
+                    upper_bounds.append(np.inf)
+                lower_bounds.append(0)
+        bounds = ((lower_bounds), (upper_bounds))
+
+    if not global_opt:
+        if 'maxfev' not in kwargs:
+            kwargs['maxfev'] = 100000
+        if 'ftol' not in kwargs:
+            kwargs['ftol'] = 1e-13
 
-    if 'maxfev' not in kwargs:
-        kwargs['maxfev'] = 100000
-    if 'ftol' not in kwargs:
-        kwargs['ftol'] = 1e-13
+        popt, pcov = curve_fit(wrapCircuit(circuit, constants), frequencies,
+                               np.hstack([Z.real, Z.imag]),
+                               p0=initial_guess, bounds=bounds, **kwargs)
 
-    popt, pcov = curve_fit(wrapCircuit(circuit, constants), f,
-                           np.hstack([Z.real, Z.imag]), p0=initial_guess,
-                           bounds=bounds, **kwargs)
+        perror = np.sqrt(np.diag(pcov))
 
-    perror = np.sqrt(np.diag(pcov))
+    else:
+        def opt_function(x):
+            """ Short function for basinhopping to optimize over.
+            We want to minimize the RMSE between the fit and the data.
+
+            Parameters
+            ----------
+            x : args
+                Parameters for optimization.
+
+            Returns
+            -------
+            function
+                Returns a function (RMSE as a function of parameters).
+            """
+            return rmse(wrapCircuit(circuit, constants)(frequencies, *x),
+                        np.hstack([Z.real, Z.imag]))
+
+        results = basinhopping(opt_function, x0=initial_guess)
+        popt = results.x
+
+        # jacobian -> covariance
+        # https://stats.stackexchange.com/q/231868
+        jac = results.lowest_optimization_result["jac"][np.newaxis]
+        try:
+            perror = inv(np.dot(jac.T, jac)) * opt_function(popt) ** 2
+        except np.linalg.LinAlgError:
+            print('Failed to compute perror')
```

Comment:
I've found that it's much easier to deal with messages like this if they are warnings (you can choose to ignore them, raise exceptions on them, etc. which can be useful, especially in automated analysis) rather than prints, so can we change this to `warnings.warn("Failed to compute perror")`? You'll also have to `import warnings` at the top of the file

Reply:
Ooh, that's a great tip! This would be a separate ticket, but I think we should change the printing behavior in `BaseCircuit.predict()` to this. 
I was recently predicting impedance within a regression and it printed "Simulating circuit based on initial parameters" like 500 times 😆

Reply:
Good call! Changed, also slightly rearranged imports per [PEP8](https://www.python.org/dev/peps/pep-0008/#imports)

Reply:
Haha, yeah @BGerwe that's exactly the printing issue that drove my original comment...

Reply:
PR to update the other print to warnings here: https://github.com/ECSHackWeek/impedance.py/pull/158